indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java

```diff
@@ -76,11 +76,34 @@ public <RetType> RetType submit(TaskAction<RetType> taskAction)
     }
 
     final long performStartTime = System.currentTimeMillis();
-    final RetType result = taskAction.perform(task, toolbox);
+    final RetType result;
+    if (isBatchAllocateAction(taskAction)) {
+      result = performBatchAllocateAction(taskAction);
+    } else {
+      result = taskAction.perform(task, toolbox);
+    }
     emitTimerMetric("task/action/run/time", taskAction, System.currentTimeMillis() - performStartTime);
     return result;
   }
 
+  private <R> boolean isBatchAllocateAction(TaskAction<R> taskAction)
+  {
+    return toolbox.canBatchSegmentAllocation()
+        && taskAction instanceof SegmentAllocateAction;
+  }
+
+  @SuppressWarnings("unchecked")
+  private <R> R performBatchAllocateAction(TaskAction<R> taskAction)
+  {
+    try {
+      SegmentAllocateAction allocateAction = (SegmentAllocateAction) taskAction;
+      return (R) allocateAction.performAsync(task, toolbox).get();
```

Comment:
The `get` call should have a timeout. Even a 10 minute timeout is fine but it shouldn't block forever.